File: backend/app/services/plaid.py
from datetime import datetime, timedelta
from typing import List, TypedDict

import plaid
from plaid.api import plaid_api
from plaid.model.link_token_create_request import LinkTokenCreateRequest
from plaid.model.link_token_create_request_user import LinkTokenCreateRequestUser
from plaid.model.products import Products
from plaid.model.country_code import CountryCode
from plaid.model.item_public_token_exchange_request import ItemPublicTokenExchangeRequest
from plaid.model.accounts_get_request import AccountsGetRequest
from plaid.model.transactions_get_request import TransactionsGetRequest
from plaid.model.transactions_get_request_options import TransactionsGetRequestOptions

from app.core.config import settings


class Transaction(TypedDict):
    id: str
    account_id: str
    amount: float
    date: datetime
    name: str
    merchant_name: str


class PlaidService:
    def __init__(self):
        # Initialize Plaid client
        configuration = plaid.Configuration(
            host=plaid.Environment.Production,  # Change to Production for live environment
            api_key={
                'clientId': settings.PLAID_CLIENT_ID,
                'secret': settings.PLAID_SECRET,
                'plaid-version': '2020-09-14'
            }
        )
        self.client = plaid_api.PlaidApi(plaid.ApiClient(configuration))


    def create_link_token(self, user_id: str) -> str:
        """
        Create a link token for initializing Plaid Link.
        
        Args:
            user_id: The unique identifier for the user
            
        Returns:
            The link token string
        """

        request = LinkTokenCreateRequest(
            user=LinkTokenCreateRequestUser(
                client_user_id=user_id
            ),
            client_name='Financeer',
            products=[Products("transactions")],
            country_codes=[CountryCode('US')],
            language='en'
        )

        response = self.client.link_token_create(request)
        return response.link_token


    def exchange_public_token(self, public_token: str) -> str:
        """
        Exchange a public token for an access token.
        
        Args:
            public_token: The public token received from Plaid Link
            
        Returns:
            The access token string
        """

        request = ItemPublicTokenExchangeRequest(
            public_token=public_token
        )

        return self.client.item_public_token_exchange(request).access_token


    def get_accounts(self, access_token: str) -> List[dict]:
        """
        Get account information for a given access token.
        
        Args:
            access_token: The access token for the item
            
        Returns:
            List of account information
        """

        request = AccountsGetRequest(access_token=access_token)

        response = self.client.accounts_get(request)

        return [
            {
                "id": account.account_id,
                "name": account.name,
                "type": account.type,
                "subtype": account.subtype,
                "mask": account.mask,
                "balances": {
                    "available": account.balances.available,
                    "current": account.balances.current,
                    "limit": account.balances.limit,
                }
            }
            for account in response.accounts
        ]


    def get_transactions(
        self,
        access_token: str,
        start_date: datetime | None = None,
        end_date: datetime | None = None,
        account_ids: list[str] | None = None,
        count: int = 100,
    ) -> list[Transaction]:
        """
        Get all transactions for a given access token between the two
        dates.

        Args:
            access_token: The access token for the item
            start_date: Start date for transactions
            end_date: End date for transactions.
            account_ids: List of account IDs to filter by.
            count: Number of transactions per API call.

        Returns:
            List of transaction data.
        """

        start_date = start_date or datetime.now() - timedelta(days=30)
        end_date = end_date or datetime.now()

        all_transactions, current_offset, total_transactions = [], 0, None
        while True:
            options = TransactionsGetRequestOptions(
                account_ids=account_ids,
                count=count,
                offset=current_offset
            )

            request = TransactionsGetRequest(
                access_token=access_token,
                start_date=start_date.date(),
                end_date=end_date.date(),
                options=options
            )
            response = self.client.transactions_get(request)

            # Store total transactions on first iteration
            if total_transactions is None:
                total_transactions = response.total_transactions

            # Add transactions from this batch
            all_transactions.extend([
                Transaction(
                    id=transaction.transaction_id,
                    account_id=transaction.account_id,
                    amount=-transaction.amount,
                    date=transaction.date,
                    name=transaction.name,
                    merchant_name=transaction.merchant_name,
                )
                for transaction in response.transactions
            ])

            # Break if we've fetched all transactions
            if len(all_transactions) >= total_transactions:
                break

            # Update offset for next batch
            current_offset += count

        return all_transactions

File: frontend/src/lib/api/plaid.js
import { api } from '@/lib/api';

import { NewLinkAccountSchema } from '@/lib/api/types';
import { ReturnAccountInfoSchema } from '@/lib/api_types';


/**
 * Creates a link token for initializing Plaid Link
 * @returns {Promise<string>} The link token
 * @throws {Error} If the API request fails
 */
export const createLinkToken = async () => {
  try {
    const { data } = await api.post('/plaid/link-token');
    return data.link_token;
  } catch (error) {
    console.error('Error creating link token:', error);
    throw new Error(error.response?.data?.detail || 'Failed to create link token');
  }
};

/**
 * Exchanges a public token for an access token
 * @param {string} publicToken The public token received from Plaid Link
 * @returns {Promise<string>} The access token
 * @throws {Error} If the API request fails
 */
export const exchangePublicToken = async (publicToken) => {
  try {
    const { data } = await api.post('/plaid/exchange-token', {
      public_token: publicToken
    });
    return data.access_token;
  } catch (error) {
    console.error('Error exchanging public token:', error);
    throw new Error(error.response?.data?.detail || 'Failed to exchange public token');
  }
};

/**
 * Gets the stored access token for the current user
 * @returns {Promise<string>} The access token
 * @throws {Error} If the API request fails
 */
export const getStoredAccessToken = async () => {
  try {
    const { data } = await api.get('/plaid/access-token');
    return data.access_token;
  } catch (error) {
    console.error('Error getting stored access token:', error);
    throw new Error(error.response?.data?.detail || 'Failed to get stored access token');
  }
};

/**
 * Gets account information from Plaid
 * @param {string | null} plaidItemId The Plaid item ID
 * @returns {Promise<ReturnAccountInfoSchema[]>} Array of account information
 * @throws {Error} If the API request fails
 */
export const getPlaidAccounts = async (plaidItemId) => {
  try {
    const { data } = await api.get('/plaid/accounts', {
      params: { plaid_item_id: plaidItemId }
    });
    return data;
  } catch (error) {
    console.error('Error getting Plaid accounts:', error);
    throw new Error(error.response?.data?.detail || 'Failed to get Plaid accounts');
  }
};

/**
 * Gets transactions from Plaid
 * @param {string} accessToken The access token for the item
 * @param {Date} startDate Start date for transactions
 * @param {Date} endDate End date for transactions
 * @param {Array<string>} accountIds Optional array of account IDs to filter by
 * @returns {Promise<Object>} Object containing transactions and total count
 * @throws {Error} If the API request fails
 */
export const getPlaidTransactions = async (accessToken, startDate, endDate, accountIds = null) => {
  try {
    const { data } = await api.get('/plaid/transactions', {
      params: {
        access_token: accessToken,
        start_date: startDate.toISOString().split('T')[0],
        end_date: endDate.toISOString().split('T')[0],
        account_ids: accountIds
      }
    });
    console.log('Received transactions data:', data);
    return data;
  } catch (error) {
    console.error('Error getting Plaid transactions:', error);
    throw new Error(error.response?.data?.detail || 'Failed to get Plaid transactions');
  }
};

/**
 * Links an existing account with a Plaid account
 * @param {NewLinkAccountSchema} linkAccountRequest The request containing the Plaid account ID, PlaidItem ID, and Account ID
 * @returns {Promise<Object>} The updated account data
 * @throws {Error} If the API request fails
 */
export const linkPlaidAccount = async (accountLink) => {
  try {
    const response = await api.post(`/plaid/link-account`, accountLink);
    return response.data;
  } catch (error) {
    console.error('Error linking Plaid account:', error);
    throw new Error(error.response?.data?.detail || 'Failed to link Plaid account');
  }
};
